tests/behat/utils/maybe-generate-wp-cli-coverage.php

```diff
@@ -2,41 +2,87 @@
-use SebastianBergmann\CodeCoverage\Report\Clover;
+use SebastianBergmann\CodeCoverage\Report\PHP as PHPReport;
 
 $root_folder = realpath( dirname( __DIR__, 3 ) );
 
+$feature   = getenv( 'BEHAT_FEATURE_TITLE' );
+$scenario  = getenv( 'BEHAT_SCENARIO_TITLE' );
+$step_line = (int) getenv( 'BEHAT_STEP_LINE' );
+$name      = "{$feature} - {$scenario} - {$step_line}";
+
+if ( empty( $feature ) || empty( $scenario ) ) {
+	return;
+}
+
 if ( ! class_exists( 'SebastianBergmann\CodeCoverage\Filter' ) ) {
 	require "{$root_folder}/vendor/autoload.php";
 }
 
+$filtered_items = new CallbackFilterIterator(
+	new DirectoryIterator( $root_folder ),
+	function ( $file ) {
+		if ( $file->isDir() && in_array( $file->getFilename(), [ 'php', 'src' ], true ) ) {
+			return true;
+		}
+
+		if ( $file->isFile() && false !== strpos( $file->getFilename(), '-command.php' ) ) {
+			return true;
+		}
+
+		return false;
+	}
+);
+
+$files = [];
+
+foreach ( $filtered_items as $item ) {
+	if ( $item->isDir() ) {
+		foreach (
+			new RecursiveIteratorIterator(
+				new RecursiveDirectoryIterator( $item->getPathname(), RecursiveDirectoryIterator::SKIP_DOTS )
+			) as $file
+		) {
+			if ( $file->isFile() && $file->getExtension() === 'php' ) {
+				$files[] = $file->getPathname();
+			}
+		}
+	} else {
+		$files[] = $item->getPathname();
+	}
+}
+
 $filter = new Filter();
-$filter->includeDirectory( "{$root_folder}/includes" );
-$filter->includeFiles( array( "{$root_folder}/plugin.php" ) );
+
+if ( method_exists( $filter, 'includeFiles' ) ) {
+	$filter->includeFiles( $files );
+} else {
+	$filter->addFilesToWhitelist( $files );
+}
 
 $coverage = new CodeCoverage(
-	( new Selector() )->forLineCoverage( $filter ),
+	class_exists( Selector::class ) ? ( new Selector() )->forLineCoverage( $filter ) : ( new Xdebug() ),
 	$filter
 );
 
-$feature  = getenv( 'BEHAT_FEATURE_TITLE' );
-$scenario = getenv( 'BEHAT_SCENARIO_TITLE' );
-$name     = "{$feature} - {$scenario}";
-
 $coverage->start( $name );
 
 register_shutdown_function(
-	static function () use ( $coverage, $feature, $scenario, $name ) {
+	static function () use ( $coverage, $feature, $scenario, $step_line, $name, $root_folder ) {
 		$coverage->stop();
 
-		$project_dir = getenv( 'BEHAT_PROJECT_DIR' );
-
 		$feature_suffix  = preg_replace( '/[^a-z0-9]+/', '-', strtolower( $feature ) );
 		$scenario_suffix = preg_replace( '/[^a-z0-9]+/', '-', strtolower( $scenario ) );
-		$filename        = "clover-behat/{$feature_suffix}-{$scenario_suffix}.xml";
-		$destination     = "{$project_dir}/build/logs/{$filename}";
+		$db_type         = strtolower( getenv( 'WP_CLI_TEST_DBTYPE' ) );
+		$destination     = "$root_folder/build/logs/$feature_suffix-$scenario_suffix-$step_line-$db_type.cov";
+
+		$dir = dirname( $destination );
+		if ( ! file_exists( $dir ) ) {
+			mkdir( $dir, 0777, true );
+		}
 
-		( new Clover() )->process( $coverage, $destination, $name );
+		( new PHPReport() )->process( $coverage, $destination );
```

Comment:
I suggest reverting this and keep using `Clover`.

What WP-CLI does for the test suite is generating separate PHP reports per scenario, and then glueing them together after all tests ran. See https://github.com/wp-cli/wp-cli-tests/blob/5d933e384003af15161099174771fbf740cc2e4a/src/Context/FeatureContext.php#L285-L310

I don't think we need this for this plugin (it's not that big).